public/app/core/components/Select/FolderPicker.tsx

```diff
@@ -82,7 +82,13 @@ export class FolderPicker extends PureComponent<Props, State> {
     await this.loadInitialValue();
   };
 
-  getOptions = async (query: string) => {
+  // when debouncing, we must use the callback form of react-select's loadOptions so we don't
+  // drop results for user input
+  loadOptions = (query: string, callback: LoadOptionsCallback<number>): void => {
+    this.searchFolders(query).then(callback);
+  };
```

Comment:
seems weird to mix promise styles. could we `async`/`await` here?

Reply:
No, we specifically cannot - that's what this whole PR is about 😅 

`await` triggers a the limitation that causes results to be dropped.

Reply:
@joshhunt are you sure? i don't see how this is any different functionally. it's just the promise syntax that's changed: 
```suggestion
  async loadOptions = (query: string, callback: LoadOptionsCallback<number>): void => {
    const options = await this.searchFolders(query);
    callback(options);
  };
```

Reply:
`async` functions return Promises. [An async function cannot return `void`](https://www.typescriptlang.org/play?#code/C4TwDgpgBAwghgGwQIzgYwNZQLxQBQAmcwcAXFAK4B2GVA9gO5UCUOAfFAG50CWBAUPzQI4AZ1FQAEhCR0oAb35RlUBHTgEA8mGA86VCbjEgqafAEcKEAE4hyo4NZ5UA5gBooaRCnQZy8JFRMZnJuPnYFJRVotH0HKGsIUTAcKDgGOB5gKAAzCGA0AAs8SxsQZgBuKOjlL0DfPETkyuqoAF8qtqA), and _returning_ a value from a debounced function is what causes the iffy behaviour.